Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -39,6 +49,276 @@
+  // string, which matches Redump not having any serials for Datel discs.
+  m_game_id = volume.GetGameTDBID();
+  if (m_game_id.size() > 4)
+    m_game_id = m_game_id.substr(0, 4);
+
+  m_revision = volume.GetRevision().value_or(0);
+  m_disc_number = volume.GetDiscNumber().value_or(0);
+  m_size = volume.GetSize();
+
+  const DiscIO::Platform volume_type = volume.GetVolumeType();
+
+  m_future = std::async(std::launch::async, [this, volume_type]() -> std::vector<PotentialMatch> {
+    std::string platform;
+    std::once_flag* download_flag;
+    DownloadStatus* download_status;
+    switch (volume_type)
+    {
+    case Platform::GameCubeDisc:
+      platform = "gc";
+      download_flag = &m_gc_download_flag;
+      download_status = &m_gc_download_status;
+      break;
+
+    case Platform::WiiDisc:
+      platform = "wii";
+      download_flag = &m_wii_download_flag;
+      download_status = &m_wii_download_status;
+      break;
+
+    default:
+      m_result.status = Status::Error;
+      return {};
+    }
+
+    std::call_once(*download_flag, DownloadDatfile, platform, download_status);
+
+    switch (*download_status)
+    {
+    case DownloadStatus::Success:
+      return ScanDatfile(ReadDatfile(platform));
+
+    case DownloadStatus::SystemNotPublic:
+      m_result = {Status::Error, Common::GetStringT("Wii data is not public yet")};
+      return {};
+
+    case DownloadStatus::Fail:
+    default:
+      m_result = {Status::Error, Common::GetStringT("Failed to connect to Redump.org")};
+      return {};
+    }
+  });
+}
+
+static std::string GetPathForPlatform(const std::string& platform)
+{
+  return File::GetUserPath(D_REDUMPCACHE_IDX) + DIR_SEP + platform + ".zip";
+}
+
+void RedumpVerifier::DownloadDatfile(const std::string& platform, DownloadStatus* status_out)
+{
+  Common::HttpRequest request;
+
+  const std::optional<std::vector<u8>> result =
+      request.Get("http://redump.org/datfile/" + platform + "/serial,version",
+                  {{"User-Agent", Common::scm_rev_str}});
+
+  if (!result)
+  {
+    *status_out = DownloadStatus::Fail;
+    return;
+  }
+
+  if (result->size() > 1 && (*result)[0] == '<' && (*result)[1] == '!')
+  {
+    // This is an HTML page, not a zip file like we want
+
+    const std::string system_not_found_error = "System \"" + platform + "\" doesn't exist.";
+    if (result->end() != std::search(result->begin(), result->end(), system_not_found_error.begin(),
+                                     system_not_found_error.end()))
+    {
+      *status_out = DownloadStatus::SystemNotPublic;
+    }
+    else
+    {
+      *status_out = DownloadStatus::Fail;
+    }
+    return;
+  }
+
+  File::IOFile(GetPathForPlatform(platform), "wb").WriteBytes(result->data(), result->size());
+}
+
+std::vector<u8> RedumpVerifier::ReadDatfile(const std::string& platform)
+{
+  unzFile file = unzOpen(GetPathForPlatform(platform).c_str());
+  Common::ScopeGuard file_guard{[&] { unzClose(file); }};
```

Comment:
I mean, we do ignore the result of `unzClose`, but should be really be calling that when `file` is `nullptr`? According to a quick google this would return a parameter error (but not blow up or anything); so it _practically_ works but is _technically_ wrong.